File: languages/csharp/Brainfuck.cs
using System;
using System.IO;

public class Program
{
    private static int[] cells = new int[0];
    private static int pos = 0;

    static void Main(string[] args)
    {
        if (args.Length != 1)
        {
            Error("You must provide a filename as an argument");
        }

        string program;
        try
        {
            program = File.ReadAllText(args[0]);
        }
        catch
        {
            Error("Could not open file");
            return;
        }

        int instr_pointer = 0;

        while (instr_pointer < program.Length)
        {
            char command = program[instr_pointer];

            switch (command)
            {
                case '+':
                    WriteCurCell(ReadCurCell() + 1);
                    break;
                case '-':
                    WriteCurCell(ReadCurCell() - 1);
                    break;
                case '>':
                    pos++;
                    break;
                case '<':
                    if (pos == 0)
                    {
                        Error("Attempted to decrement data pointer below 0");
                    }
                    pos--;
                    break;
                case '.':
                    Console.Write((char)ReadCurCell());
                    break;
                case ',':
                    WriteCurCell(Console.Read());
                    break;
                case '[':
                    if (ReadCurCell() == 0)
                    {
                        int depth = 1;
                        while (depth > 0 && instr_pointer < program.Length)
                        {
                            instr_pointer++;
                            if (program[instr_pointer] == '[')
                            {
                                depth++;
                            }
                            else if (program[instr_pointer] == ']')
                            {
                                depth--;
                            }
                        }
                    }
                    break;
                case ']':
                    if (ReadCurCell() != 0)
                    {
                        int depth = 1;
                        while (depth > 0 && instr_pointer >= 0)
                        {
                            instr_pointer--;
                            if (program[instr_pointer] == ']')
                            {
                                depth++;
                            }
                            else if (program[instr_pointer] == '[')
                            {
                                depth--;
                            }
                        }
                    }
                    break;
            }

            instr_pointer++;
        }
    }

    static int ReadCurCell()
    {
        if (pos >= cells.Length)
        {
            return 0;
        }
        else
        {
            return cells[pos];
        }
    }

    static void WriteCurCell(int value)
    {
        if (pos >= cells.Length)
        {
            Array.Resize(ref cells, pos + 1);
        }

        cells[pos] = value;
    }

    static void Error(string msg)
    {
        Console.Error.WriteLine(msg);
        Environment.Exit(1);
    }
}

File: languages/java/Brainfuck.java

import java.nio.file.*;
import java.io.*;
import java.util.*;

public class Brainfuck {
    static class State {
        ArrayList<Integer> cells = new ArrayList<>();
        int pos = 0;

        int readCurCell() {
            if (this.pos >= this.cells.size()) {
                return 0;
            } else {
                return this.cells.get(this.pos);
            }
        }

        void writeCurCell(int value) {
            while (this.pos >= this.cells.size()) {
                this.cells.add(0);
            }

            this.cells.set(this.pos, value);
        }

        void incCurCell() {
            this.writeCurCell(this.readCurCell() + 1);
        }

        void decCurCell() {
            this.writeCurCell(this.readCurCell() - 1);
        }
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.err.println("You must provide a filename as an argument");
            System.exit(1);
        }

        String program = Files.readString(Path.of(args[0]));

        State state = new State();

        int instrPointer = 0;

        while (instrPointer < program.length()) {
            char command = program.charAt(instrPointer);

            switch (command) {
                case '+':
                    state.incCurCell();
                    break;
                case '-':
                    state.decCurCell();
                    break;
                case '>':
                    state.pos++;
                    break;
                case '<':
                    if (state.pos == 0) {
                        System.err.println("Attempted to decrement data pointer below 0 at position " + instrPointer);
                        System.exit(1);
                    }
                    state.pos--;
                    break;
                case '.':
                    System.out.print((char) state.readCurCell());
                    break;
                case ',':
                    throw new UnsupportedOperationException("',' command is not supported");
                case '[':
                    if (state.readCurCell() == 0) {
                        int depth = 1;
                        int pos = instrPointer;

                        while (depth > 0) {
                            pos++;

                            if (program.charAt(pos) == '[') {
                                depth++;
                            } else if (program.charAt(pos) == ']') {
                                depth--;
                            }

                            if (depth > 0 && pos == program.length()) {
                                System.err.println("Unmatched '[' at position " + instrPointer);
                                System.exit(1);
                            }
                        }

                        instrPointer = pos;
                    }
                    break;
                case ']':
                    if (state.readCurCell() != 0) {
                        int depth = 1;
                        int pos = instrPointer;

                        while (depth > 0) {
                            pos--;

                            if (program.charAt(pos) == ']') {
                                depth++;
                            } else if (program.charAt(pos) == '[') {
                                depth--;
                            }

                            if (depth > 0 && pos < 0) {
                                System.err.println("Unmatched ']' at position " + instrPointer);
                                System.exit(1);
                            }
                        }

                        instrPointer = pos;
                    }
                    break;
            }

            instrPointer++;
        }
    }
}
